amp/parse.go

```diff
@@ -32,6 +41,114 @@ type Size struct {
 	Width          int64
 }
 
+// Policy consent types
+const (
+	ConsentNone      = 0
+	ConsentTCF1      = 1
+	ConsentTCF2      = 2
+	ConsentUSPrivacy = 3
+)
+
+// ReadPolicy returns a privacy writer in accordance to the query values consent, consent_type and gdpr_applies.
+// Returned policy writer could either be GDPR, CCPA or NilPolicy. The second return value is a warning.
+func ReadPolicy(ampParams Params, pbsConfigGDPREnabled bool) (privacy.PolicyWriter, error) {
+	if len(ampParams.Consent) == 0 {
+		return privacy.NilPolicyWriter{}, nil
+	}
+
+	var rv privacy.PolicyWriter = privacy.NilPolicyWriter{}
+	var warning error
+	var warningMsg string
+
+	// If consent_type was set to CCPA or GDPR TCF2, we return a valid writer even if the consent string is invalid
+	switch ampParams.ConsentType {
+	case ConsentTCF1:
+		warningMsg = "TCF1 consent is deprecated and no longer supported."
+	case ConsentTCF2:
+		if pbsConfigGDPREnabled {
+			rv = buildGdprTCF2ConsentWriter(ampParams)
+			// Log warning if GDPR consent string is invalid
+			warningMsg = validateTCf2ConsentString(ampParams.Consent)
+		}
+	case ConsentUSPrivacy:
+		rv = ccpa.ConsentWriter{ampParams.Consent}
+		if ccpa.ValidateConsent(ampParams.Consent) {
+			if parseGdprApplies(ampParams.GdprApplies) == 1 {
```

Comment:
I was pondering, since gdpr is opt out and requires careful tracking of consent, do we want to default to no consent in case of this mismatch?

Reply:
I think we should. Would make perfect sense. Should I still log the warning?

Reply:
I spoke with Julien Delhommeau and he suggests that `consent_type` is reasonable to take as authoritative as to which consent framework applies to the request. So we are good with the current code.